sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp

```diff
@@ -4,256 +4,443 @@
+      sycl::sub_group,
+      typename std::enable_if_t<Layout == matrix_layout::row_major ||
+                                Layout == matrix_layout::col_major>> {
+    float data[8];
+  };
+
+  } // namespace experimental::matrix
+
+  namespace detail {
+  using namespace experimental;
+
+  template <typename U, matrix::matrix_type T, matrix::matrix_use MT,
+            size_t NumRows, size_t NumCols, matrix::matrix_layout Layout,
+            access::address_space Space, typename Cond = void>
+  struct joint_matrix_load_impl {
+    void load(matrix::joint_matrix<T, MT, NumRows, NumCols, Layout> &res,
+              multi_ptr<U, Space> src, size_t stride);
+  };
+
+  template <matrix::matrix_layout Layout> constexpr int get_layout_id();
+
+  template <> constexpr int get_layout_id<matrix::matrix_layout::row_major>() {
+    return 0;
+  }
+
+  template <> constexpr int get_layout_id<matrix::matrix_layout::col_major>() {
+    return 1;
+  }
+
+  template <matrix::matrix_layout Layout, access::address_space Space>
+  struct joint_matrix_load_impl<
+      double, matrix::matrix_type::fp64, matrix::matrix_use::a, 8, 4, Layout,
+      Space,
+      typename std::enable_if_t<Layout == matrix::matrix_layout::row_major ||
+                                Layout == matrix::matrix_layout::col_major>> {
+    void load(matrix::joint_matrix<matrix::matrix_type::fp64,
+                                   matrix::matrix_use::a, 8, 4, Layout> &res,
+              multi_ptr<double, Space> src, size_t stride) {
 
 #ifdef __NVPTX__
 #ifdef __SYCL_DEVICE_ONLY__
-    __dmma_m8n8k4_ld_a(res.data, src.get(), stride, get_layout_id<Layout>());
+      __dmma_m8n8k4_ld_a(res.data, src.get(), stride, get_layout_id<Layout>());
 #endif
 #endif
-  }
-};
+    }
+  };
+
+  template <matrix::matrix_layout Layout, access::address_space Space>
+  struct joint_matrix_load_impl<
+      float, matrix::matrix_type::bf16, matrix::matrix_use::a, 16, 16, Layout,
+      Space,
+      typename std::enable_if_t<Layout == matrix::matrix_layout::row_major ||
+                                Layout == matrix::matrix_layout::col_major>> {
+    void load(matrix::joint_matrix<matrix::matrix_type::bf16,
+                                   matrix::matrix_use::a, 16, 16, Layout> &res,
+              multi_ptr<float, Space> src, size_t stride) {
 
-template <matrix::matrix_layout Layout, access::address_space Space>
-struct joint_matrix_load_impl<
-    double, matrix::matrix_use::b, 4, 8, Layout, Space,
-    typename std::enable_if_t<Layout == matrix::matrix_layout::row_major ||
-                              Layout == matrix::matrix_layout::col_major>> {
-  void
-  load(matrix::joint_matrix<double, matrix::matrix_use::b, 4, 8, Layout> &res,
-       multi_ptr<double, Space> src, size_t stride) {
 #ifdef __NVPTX__
 #ifdef __SYCL_DEVICE_ONLY__
-    __dmma_m8n8k4_ld_b(res.data, src.get(), stride, get_layout_id<Layout>());
+      uint32_t val[128];
```

Comment:
A SPIRV function will be utilized here to allow a more efficient conversion from the fp32 to bf16 formats. Currently the threads repeatedly copy/convert the same data.  This current version is meant as a simple proof of concept example only.

Reply:
I am adding @AlexeySotkin here because I know he has been working on conversion functions for these types.

Reply:
Thanks, I've just looked at the bf16 conversion extension proposal.  
It is worth mentioning that for the joint_matrix_mad operations tensor cores required that two bf16 floats are packed into a single uint32_t.  I believe there only currently exists an operation to pack/convert two fp32's into a single uint32_t - It doesn't appear to be possible to directly pack two bf16's stored in uint16_t's into a single uint32_t.

Reply:
Can you send the spec that explains this packing requirement?
This seems similar to the VNNI packing requirement for AMX/DPAS (packing of low precision to 32 bits)

Reply:
Sure.  The spec is spread out a little:

Firstly under the header "Floating point fragments for .bf16 data format" in this [section](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#warp-level-matrix-fragment) it states that for the A,B matrices the data is stored via:

"A vector expression of four .b32 registers, with each register containing two elements from the matrix."

The conversion/packing information can be found [here](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cvt).